### ISSUE TITLE
BuildVectorType with a dependent array type is crashing the compiler

### ISSUE THREAD

**Issue by zahiraam** [llvm/llvm-project#46886]
|  |  |
| --- | --- |
| Bugzilla Link | [47542](https://llvm.org/bz47542) |
| Version | unspecified |
| OS | Windows NT |
| CC | @DougGregor,@zygoloid,@zahiraam |

## Extended Description 
Please see

https://godbolt.org/z/MbGn68

for illustration.

**Comment by wheatman** [comment]
If this is referring to an out of memory issue that happened with 11.0.1, then it was fixed for 12.0.0
https://godbolt.org/z/zn46fjn7W

The code was 
```
template<typename T> struct A
{
    int x[sizeof(T)] __attribute__((vector_size(8))); 
};
```

The dump was 
```
LLVM ERROR: out of memory
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.	Program arguments: /opt/compiler-explorer/clang-11.0.1/bin/clang++ -gdwarf-4 -g -o /app/output.s -mllvm --x86-asm-syntax=intel -S --gcc-toolchain=/opt/compiler-explorer/gcc-10.2.0 -fcolor-diagnostics -fno-crash-diagnostics <source> 
1.	<source>:3:53: current parser token ';'
2.	<source>:1:22: parsing struct/union/class body 'A'
 #0 0x0000560af62acc5a llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x2cacc5a)
 #1 0x0000560af62aaa04 llvm::sys::RunSignalHandlers() (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x2caaa04)
 #2 0x0000560af62aac75 llvm::sys::CleanupOnSignal(unsigned long) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x2caac75)
 #3 0x0000560af621f4a0 CrashRecoverySignalHandler(int) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x2c1f4a0)
 #4 0x00007fca26184420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #5 0x00007fca25c5100b raise (/lib/x86_64-linux-gnu/libc.so.6+0x4300b)
 #6 0x00007fca25c30859 abort (/lib/x86_64-linux-gnu/libc.so.6+0x22859)
 #7 0x0000560af6226261 llvm::report_bad_alloc_error(char const*, bool) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x2c26261)
 #8 0x0000560af6226295 (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x2c26295)
 #9 0x00007fca26016b50 operator new(unsigned long) (/lib/x86_64-linux-gnu/libstdc++.so.6+0xaab50)
#10 0x00007fca26016b6d operator new(unsigned long, std::nothrow_t const&) (/lib/x86_64-linux-gnu/libstdc++.so.6+0xaab6d)
#11 0x0000560af8585c80 GetTypeSourceInfoForDeclarator((anonymous namespace)::TypeProcessingState&, clang::QualType, clang::TypeSourceInfo*) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x4f85c80)
#12 0x0000560af858b9fc GetFullTypeForDeclarator((anonymous namespace)::TypeProcessingState&, clang::QualType, clang::TypeSourceInfo*) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x4f8b9fc)
#13 0x0000560af858e304 clang::Sema::GetTypeForDeclarator(clang::Declarator&, clang::Scope*) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x4f8e304)
#14 0x0000560af800002d clang::Sema::HandleField(clang::Scope*, clang::RecordDecl*, clang::SourceLocation, clang::Declarator&, clang::Expr*, clang::InClassInitStyle, clang::AccessSpecifier) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x4a0002d)
#15 0x0000560af80a37ed clang::Sema::ActOnCXXMemberDeclarator(clang::Scope*, clang::AccessSpecifier, clang::Declarator&, llvm::MutableArrayRef<clang::TemplateParameterList*>, clang::Expr*, clang::VirtSpecifiers const&, clang::InClassInitStyle) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x4aa37ed)
#16 0x0000560af7e262cf clang::Parser::ParseCXXClassMemberDeclaration(clang::AccessSpecifier, clang::ParsedAttributes&, clang::Parser::ParsedTemplateInfo const&, clang::ParsingDeclRAIIObject*) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x48262cf)
#17 0x0000560af7e2889b clang::Parser::ParseCXXClassMemberDeclarationWithPragmas(clang::AccessSpecifier&, clang::Parser::ParsedAttributesWithRange&, clang::TypeSpecifierType, clang::Decl*) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x482889b)
#18 0x0000560af7e290a1 clang::Parser::ParseCXXMemberSpecification(clang::SourceLocation, clang::SourceLocation, clang::Parser::ParsedAttributesWithRange&, unsigned int, clang::Decl*) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x48290a1)
#19 0x0000560af7e2aee3 clang::Parser::ParseClassSpecifier(clang::tok::TokenKind, clang::SourceLocation, clang::DeclSpec&, clang::Parser::ParsedTemplateInfo const&, clang::AccessSpecifier, bool, clang::Parser::DeclSpecContext, clang::Parser::ParsedAttributesWithRange&) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x482aee3)
#20 0x0000560af7e0bc5b clang::Parser::ParseDeclarationSpecifiers(clang::DeclSpec&, clang::Parser::ParsedTemplateInfo const&, clang::AccessSpecifier, clang::Parser::DeclSpecContext, clang::Parser::LateParsedAttrList*) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x480bc5b)
#21 0x0000560af7e9b987 clang::Parser::ParseSingleDeclarationAfterTemplate(clang::DeclaratorContext, clang::Parser::ParsedTemplateInfo const&, clang::ParsingDeclRAIIObject&, clang::SourceLocation&, clang::ParsedAttributes&, clang::AccessSpecifier) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x489b987)
#22 0x0000560af7e9e3dc clang::Parser::ParseTemplateDeclarationOrSpecialization(clang::DeclaratorContext, clang::SourceLocation&, clang::ParsedAttributes&, clang::AccessSpecifier) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x489e3dc)
#23 0x0000560af7e9e834 clang::Parser::ParseDeclarationStartingWithTemplate(clang::DeclaratorContext, clang::SourceLocation&, clang::ParsedAttributes&, clang::AccessSpecifier) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x489e834)
#24 0x0000560af7e132d5 clang::Parser::ParseDeclaration(clang::DeclaratorContext, clang::SourceLocation&, clang::Parser::ParsedAttributesWithRange&, clang::SourceLocation*) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x48132d5)
#25 0x0000560af7deeb9b clang::Parser::ParseExternalDeclaration(clang::Parser::ParsedAttributesWithRange&, clang::ParsingDeclSpec*) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x47eeb9b)
#26 0x0000560af7df0289 clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, bool) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x47f0289)
#27 0x0000560af7df08e7 clang::Parser::ParseFirstTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x47f08e7)
#28 0x0000560af7de3e93 clang::ParseAST(clang::Sema&, bool, bool) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x47e3e93)
#29 0x0000560af6a98809 clang::FrontendAction::Execute() (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x3498809)
#30 0x0000560af6a5232b clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x345232b)
#31 0x0000560af6b5ee9b clang::ExecuteCompilerInvocation(clang::CompilerInstance*) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x355ee9b)
#32 0x0000560af44867ac cc1_main(llvm::ArrayRef<char const*>, char const*, void*) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0xe867ac)
#33 0x0000560af44832ad ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0xe832ad)
#34 0x0000560af692a055 void llvm::function_ref<void ()>::callback_fn<clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*) const::'lambda'()>(long) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x332a055)
#35 0x0000560af621f583 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x2c1f583)
#36 0x0000560af692ab30 clang::driver::CC1Command::Execute(llvm::ArrayRef<llvm::Optional<llvm::StringRef> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*) const (.part.148) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x332ab30)
#37 0x0000560af6904d05 clang::driver::Compilation::ExecuteCommand(clang::driver::Command const&, clang::driver::Command const*&) const (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x3304d05)
#38 0x0000560af690574f clang::driver::Compilation::ExecuteJobs(clang::driver::JobList const&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) const (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x330574f)
#39 0x0000560af690e2e5 clang::driver::Driver::ExecuteCompilation(clang::driver::Compilation&, llvm::SmallVectorImpl<std::pair<int, clang::driver::Command const*> >&) (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0x330e2e5)
#40 0x0000560af43c3421 main (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0xdc3421)
#41 0x00007fca25c32083 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24083)
#42 0x0000560af4482e2a _start (/opt/compiler-explorer/clang-11.0.1/bin/clang+++0xe82e2a)
clang-11: error: clang frontend command failed due to signal (use -v to see invocation)
Compiler returned: 254
```

However, this code still doesn't compile it gives an error while gcc compiles the code fine
```
<source>:3:37: error: invalid vector element type 'int [sizeof(T)]'
    int x[sizeof(T)] __attribute__((vector_size(8))); 
                                    ^
1 error generated.
Compiler returned: 1
```

**Comment by Endilll** [comment]
@wheatman added "crash-on-invalid" label
